src/node_i18n.cc

```diff
@@ -601,6 +602,13 @@ void SetDefaultTimeZone(const char* tzid) {
   CHECK(U_SUCCESS(status));
 }
 
+void SetDefaultLocale(const char* localeid) {
+  UErrorCode status = U_ZERO_ERROR;
+
+  uloc_setDefault(localeid, &status);
+  CHECK(U_SUCCESS(status));
```

Comment:
Does this CHECK when I pass in a bad locale? That's not very user-friendly.

Reply:
If you pass an invalid locale, it still pass but the locale will be set to 'und'. I assume 'und' stands for 'undefined'.

If you have a proposal on how to implement it better, i will consider and implement it.

I thought of maybe do a getDefaultLocale first and store it temporary and if we get 'und' reset it to the temprorary stored value

Reply:
Oh, you're right. Now that you mention it: uloc_setDefault() never fails, it just ignores errors, so the CHECK is an appropriate just-in-case...

...but having said that, I'd say it's important to print a warning when something goes wrong. You can use the C++ API from unicode/locid.h to do that:
```c++
icu::Locale locale(localeid);
if (locale.isBogus()) {
  // bad localeid
} else {
  UErrorCode status = U_ZERO_ERROR;
  icu::Locale::setDefault(locale, &status);  // equivalent to uloc_setDefault()
  if (!U_SUCCESS(status)) {
    // handle error
  }
}
```

Reply:
I tried that yesterday, but it seems that it does not mark an invalid locale as bogus.

I did 

```cpp
void SetDefaultLocale(const char* localeid) {
  UErrorCode status = U_ZERO_ERROR;

  icu::Locale icuLocale(localeid);

  if (!icuLocale.isBogus()) {
    printf("not bogus");
    uloc_setDefault(localeid, &status);
    CHECK(U_SUCCESS(status));
  }
}
```

And `und` is "not bogus".

Reply:
test would be 

```js
spawnSyncAndExitWithoutError(
  process.execPath,
  [
    '--locale=ääö',
    '-p',
    'Intl?.Collator().resolvedOptions().locale',
  ],
  {
    signal: null,
    stdout: RegExp(defaultLocale),
  });
  
```

Reply:
Oh :facepalm:

I guess you could check if the locale is a known locale, like this:
```c++
icu::Locale icuLocale(localeid);
int32_t count;
icu::Locale* begin = icu::Locale::getAvailableLocales(count);
icu::Locale* end = &begin[count];
bool found = end != std::find(begin, end, locale);
```

Reply:
Thats weird. I implement it to your suggestion. But then the japanese unit test fails. Despite it would work without the change? chinese (`zh`) and korean (`ko`) pass, but japanese (`jp-JP` or `jp`) not.

```c++
void SetDefaultLocale(const char* localeid) {
  UErrorCode status = U_ZERO_ERROR;
  icu::Locale locale(localeid);
  int32_t numLocales = 0;
  const icu::Locale* begin = icu::Locale::getAvailableLocales(numLocales);
  const icu::Locale* end = &begin[numLocales];
  bool found = end != std::find(begin, end, locale);

  if (found) {
    uloc_setDefault(localeid, &status);
    CHECK(U_SUCCESS(status));
  }
}
```

Reply:
Okay, I throw my hands up. I guess you should check for "und" afterwards then.

Reply:
Man, this is annoying. When I use directly call getDefaultLocale I get not 'und'. But if I call `Intl?.Collator().resolvedOptions().locale` I get 'und'. So confusing.

Reply:
Japanese is `ja` not `jp`. You could check for und _unless the user explicitly set the locale to und_.

und just means the root locale, unknown language.